simpleclient_spring_web/src/main/java/io/prometheus/client/spring/web/MethodTimer.java

```diff
@@ -0,0 +1,96 @@
+package io.prometheus.client.spring.web;
+
+import io.prometheus.client.Summary;
+import org.apache.commons.lang3.StringUtils;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+
+import java.util.HashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * This class automatically times (via aspectj) the execution of methods by their signature, if it's been enabled via {@link EnablePrometheusTiming}
+ * and in methods annotated (or within classes annotated) with {@link PrometheusTimeMethods}
+ *
+ * @author Andrew Stuart
+ */
+@Aspect
+@ControllerAdvice
+@Component
+public class MethodTimer {
+    public static final String METRIC_NAME = "prometheus_method_timing_seconds";
+
+    public static final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    public static final HashMap<String, Summary> collectors = new HashMap<String, Summary>(10);
+
+    public static final Summary defaultSummary = Summary.build()
+            .name(METRIC_NAME)
+            .help("Automatic annotation-driven method timing")
+            .labelNames("signature")
+            .register();
+
+    private Summary getSummary(PrometheusTimeMethods annot) {
+        Summary summary;
+        String name = annot.value();
+
+        // Try to read first
+        try {
+            lock.readLock().lock();
+            if (collectors.containsKey(name)) {
+                return collectors.get(name);
+            }
+        } finally {
+            lock.readLock().unlock();
+        }
+        try {
+            lock.writeLock().lock();
+
+            // no readers can exist here, so we check again in case multiple thread were waiting on the write lock
+            // and one of them got here first
+            if (collectors.containsKey(name)) {
+                return collectors.get(name);
+            }
+
+            // Only one thread may get here, as writeLock is fully mutually exclusive
+            summary = Summary.build()
+                    .name(name)
+                    .help(annot.help())
+                    .labelNames("signature")
```

Comment:
Well, for one, if it's performant enough I don't see why profiling in production would be a bad thing. Getting that level of visibility into your application's real performance, and the ability to drill down, seems like a win. But that said, my goal was really just to make it easy to set up whatever level of granularity users want, and still provide meaningful output.

In my use case, it's great to have whole spring `@Controller` and `@Service` classes annotated and get metrics on individual method performance, at least at those levels.